src/bluesearch/database/topic_rule.py

```diff
@@ -0,0 +1,125 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Implementation of the TopicRule data structure."""
+from __future__ import annotations
+
+import re
+from typing import Iterable, Optional
+
+import pydantic
+
+from bluesearch.database.article import ArticleSource
+from bluesearch.database.topic_info import TopicInfo
+
+
+class TopicRule(pydantic.BaseModel, arbitrary_types_allowed=True):
+    """The topic rule."""
+
+    # None always represent wildcards
+    level: Optional[str] = None  # "article" or "journal"
+    source: Optional[ArticleSource] = None  # "arxiv", ... , "pubmed"
+    pattern: Optional[re.Pattern] = None  # regex pattern to match
+
+    @pydantic.validator("level")
+    def check_level_value(cls, value):
+        """Check the level parameter value."""
+        if value is not None and value not in {"article", "journal"}:
+            raise ValueError(f"Unsupported level {value}")
+        return value
+
+    @pydantic.validator("source", pre=True)
+    def convert_source_to_article_source(cls, value) -> ArticleSource | None:
+        """Check the source parameter value."""
+        if value is not None:
+            try:
+                value = ArticleSource(value)
+            except ValueError:
+                raise ValueError(f"Unsupported source {value}") from None
+
+        return value
+
+    @pydantic.validator("pattern", pre=True)
+    def convert_pattern_to_re(cls, value) -> re.Pattern | None:
+        """Check the pattern parameter value."""
+        if value is not None:
+            try:
+                value = re.compile(value)
+            except re.error:
+                raise ValueError(f"Unsupported pattern {value}") from None
+
+        return value
+
+    def match(self, topic_info: TopicInfo) -> bool:
+        """Determine whether a topic_info matches the rule."""
+        # Source
+        if self.source is not None and self.source is not topic_info.source:
+            return False
+
+        if self.pattern is None:
+            return True
+
+        if self.level is None or self.level == "article":
+            for topic_list in topic_info.article_topics.values():
+                if any(self.pattern.search(topic) for topic in topic_list):
+                    return True
+
+        if self.level is None or self.level == "journal":
+            for topic_list in topic_info.journal_topics.values():
+                if any(self.pattern.search(topic) for topic in topic_list):
+                    return True
+
+        return False
+
+
+def check_accepted(
```

Comment:
IMO this would introduce circular imports between `topic_rule.py` and `topic_info.py`
* `TopicRule.match` inputs `TopicInfo`
* `TopicInfo.check_topic_rules` inputs `list[TopicRule]`

Reply:
Regarding the name, sure, I will change it!

Reply:
Done in fe748a5